### PR TITLE
[NU-1836] Add list methods as extension methods to array

### PR DESCRIPTION
**Pull request by lukasz-bigorajski**
## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

**Review by ZbyszekMM — COMMENTED** [review]
The docs part looks good to me - the casting section is adjacent to 'Type conversions' section. 

Pls note, that I was to able to review the Java part.

File: scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ExtensionMethods.scala [hunk 1]
```diff
@@ -40,15 +43,22 @@ object ExtensionMethods {
 }
 
 trait ExtensionMethodsFactory {
-  def create(target: Any, classLoader: ClassLoader, set: ClassDefinitionSet): Any
+  // Factory method should be as easy and lightweight as possible because it's fired with every method execution.
```
Comment (arkadius) [inline scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ExtensionMethods.scala:46]: This is very important comment. You are right that it looks confusing that we have a factory that is used very often. I did some experiment after which IMO it became a little less tricky: https://github.com/TouK/nussknacker/pull/7045/files

File: scenario-compiler/src/main/java/pl/touk/nussknacker/engine/spel/internal/ExtensionAwareMethodsDiscovery.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class ExtensionAwareMethodsDiscovery {
```
Comment (arkadius) [inline scenario-compiler/src/main/java/pl/touk/nussknacker/engine/spel/internal/ExtensionAwareMethodsDiscovery.java:16]: Why is it in java? Can you you left some comment?

**Review by arkadius — COMMENTED** [review]
Looks good, I added some comments

File: scenario-compiler/src/main/java/pl/touk/nussknacker/engine/spel/internal/ExtensionAwareMethodsDiscovery.java (hunk 2 above)
Comment (lukasz-bigorajski) [reply scenario-compiler/src/main/java/pl/touk/nussknacker/engine/spel/internal/ExtensionAwareMethodsDiscovery.java]: Moved to scala

File: scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ExtensionMethods.scala [hunk 3]
```diff
@@ -23,8 +25,9 @@ class ExtensionMethodsInvoker(classLoader: ClassLoader, classDefinitionSet: Clas
 
 object ExtensionMethods {
 
-  private[extension] val extensions = List(
-    new Extension(classOf[Cast], Cast, Cast)
+  val extensions = List(
+    new Extension(classOf[Cast], Cast, Cast, Cast),
```
[the remark this answers is not in the document]
Comment (lukasz-bigorajski) [reply scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ExtensionMethods.scala]: Done

File: scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ExtensionMethods.scala (hunk 1 above)
Comment (lukasz-bigorajski) [reply scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ExtensionMethods.scala]: Done

**Review by arkadius — APPROVED** [review]

**Pull request merged**